File: jj/programmers/level1/fruit_seller/fruit_seller.py
# source: https://school.programmers.co.kr/learn/courses/30/lessons/135808?language=python3

'''
사과의 상태: 1점부터 k점까지

한 상자의 가격 결정 기준
- 한 상자에 사과 m개씩 담아 포장
- 가장 낮은 점수가 p점인 경우, 사과 한 상자의 가격은 p * m

가능한 많은 사과를 팔았을 때, 얻을 수 있는 최대 이익 계산
상자 단위로만 판매, 남는 사과는 버린다.
'''

def solution(k, m, score):
    answer = 0
    
    score.sort(reverse=True)
    
    max_num_for_sale = len(score) // m * m
    
    for i in range(0, len(score), m):
        if i + m <= len(score):
            answer += min(score[i:i + m]) * m
    
    return answer

File: jj/programmers/level1/fruit_seller/fruit_seller.java
// source: https://school.programmers.co.kr/learn/courses/30/lessons/135808?language=java

import java.util.Arrays;
import java.util.Collections;

class Solution {
    public int getNumForSale(Integer[] score, int m) {

        return score.length / m * m;
    }

    public int getMinScore(
        int m,
        Integer[] score,
        int index
    ) {
        int minScore = 1_000_000;
        
        for (int i = index; i < index + m; i++) {
            minScore = Math.min(minScore, score[i]);
        }
        
        return minScore;
    }

    public int computeProfit(
        int numForSale,
        int m,
        Integer[] score
    ) {
        int profit = 0;

        for (int i = 0; i < numForSale; i += m) {
            profit += getMinScore(m, score, i) * m;
        }

        return profit;
    }

    public int solution(int k, int m, int[] score) {
        Integer[] rank = Arrays.stream(score).boxed().toArray(Integer[]::new);
        Arrays.sort(rank, Collections.reverseOrder());

        int numForSale = getNumForSale(rank, m);

        return computeProfit(numForSale, m, rank);
    }
}